File: sdks/dotnet/Vexil/EnvProvider.cs
namespace Vexil;

/// <summary>
/// Reads flags from environment variables.
/// Flag name "dark-mode" maps to env var "FLAG_DARK_MODE".
/// </summary>
public class EnvProvider : IFlagProvider
{
    private static string FlagNameToEnvVar(string name)
        => "FLAG_" + name.ToUpperInvariant().Replace('-', '_');

    public Flag? GetFlag(string name)
    {
        var envName = FlagNameToEnvVar(name);
        var value = Environment.GetEnvironmentVariable(envName);
        if (value is null) return null;

        return new Flag
        {
            Name = name,
            Type = "string",
            Value = value
        };
    }

    public IReadOnlyList<Flag> GetAllFlags()
    {
        var flags = new List<Flag>();
        foreach (var entry in Environment.GetEnvironmentVariables())
        {
            if (entry is System.Collections.DictionaryEntry de
                && de.Key is string key
                && de.Value is string val
                && key.StartsWith("FLAG_"))
            {
                var name = key[5..].ToLowerInvariant().Replace('_', '-');
                flags.Add(new Flag { Name = name, Type = "string", Value = val });
            }
        }
        return flags;
    }

    public void Dispose() { }
}

File: sdks/go/vexil/env_provider.go
package vexil

import (
	"fmt"
	"os"
	"strings"
)

// envProvider reads flags from environment variables.
// Flag name "dark-mode" maps to env var "FLAG_DARK_MODE".
type envProvider struct{}

func flagNameToEnvVar(name string) string {
	return "FLAG_" + strings.ToUpper(strings.ReplaceAll(name, "-", "_"))
}

func (p *envProvider) GetFlag(name string) (Flag, error) {
	envName := flagNameToEnvVar(name)
	val, ok := os.LookupEnv(envName)
	if !ok {
		return Flag{}, fmt.Errorf("flag %s not found (env var %s)", name, envName)
	}
	return Flag{
		Name:  name,
		Type:  "string",
		Value: val,
	}, nil
}

func (p *envProvider) GetAllFlags() ([]Flag, error) {
	var flags []Flag
	for _, env := range os.Environ() {
		parts := strings.SplitN(env, "=", 2)
		if len(parts) != 2 {
			continue
		}
		key, val := parts[0], parts[1]
		if !strings.HasPrefix(key, "FLAG_") {
			continue
		}
		name := strings.ToLower(strings.ReplaceAll(strings.TrimPrefix(key, "FLAG_"), "_", "-"))
		flags = append(flags, Flag{
			Name:  name,
			Type:  "string",
			Value: val,
		})
	}
	return flags, nil
}

func (p *envProvider) Close() error { return nil }
